File: src/workspace/.openhands/skills/verify/scripts/verify.py
#!/usr/bin/env python3
"""Verify a migrated Manifest V3 Chrome extension.

Launches Chromium (the one bundled in the container) with the extension loaded, waits
for its service worker to register, exercises the extension's pages, and captures
runtime errors from:
  - the extension service worker (Runtime.exceptionThrown, console error/warning)
  - the extension's own pages (popup/options): console errors + uncaught page errors

Playwright is used to launch and drive the browser; the service worker is observed via
the Chrome DevTools Protocol over the remote-debugging port (Playwright's high-level
service_workers list is unreliable for event-driven MV3 workers).

Usage:
    verify.py <extension-dir> [report-path]

Writes a JSON report and exits non-zero if the extension failed to load or any errors
were captured.
"""

import json
import os
import shutil
import sys
import time
import urllib.request

from playwright.sync_api import sync_playwright

try:
    import websocket  # websocket-client
except ImportError:  # pragma: no cover - dependency is installed by provisioning
    websocket = None

DEBUG_PORT = 9222
LOAD_TIMEOUT_S = 20
COLLECT_S = 5

_CHROME_CANDIDATES = (
    os.environ.get("CHROME_BIN"),
    "/usr/bin/chromium",
    "/usr/bin/chromium-browser",
    "/usr/bin/google-chrome",
)


def find_chrome() -> str:
    for cand in _CHROME_CANDIDATES:
        if cand and (os.path.isfile(cand) or shutil.which(cand)):
            return cand
    raise SystemExit("No Chromium/Chrome binary found in the container.")


def http_json(path: str):
    with urllib.request.urlopen(f"http://127.0.0.1:{DEBUG_PORT}{path}", timeout=5) as r:
        return json.loads(r.read())


def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: verify.py <extension-dir> [report-path]", file=sys.stderr)
        return 2

    extension_dir = os.path.abspath(sys.argv[1])
    report_path = os.path.abspath(sys.argv[2]) if len(sys.argv) > 2 else "/workspace/test_report.json"
    chrome = find_chrome()

    errors: list[dict] = []
    warnings: list[dict] = []
    logs: list[dict] = []

    def record(bucket, source, text):
        if text:
            bucket.append({"source": source, "text": str(text)})

    report = {
        "chromeVersion": None,
        "extensionId": None,
        "loaded": False,
        "errors": errors,
        "warnings": warnings,
        "logs": logs,
    }

    user_data_dir = "/tmp/verify-profile"
    shutil.rmtree(user_data_dir, ignore_errors=True)

    with sync_playwright() as p:
        context = p.chromium.launch_persistent_context(
            user_data_dir=user_data_dir,
            executable_path=chrome,
            headless=False,  # extensions require non-headless or --headless=new
            args=[
                "--headless=new",
                f"--remote-debugging-port={DEBUG_PORT}",
                # Allow the local CDP WebSocket connection (Chrome blocks it otherwise).
                f"--remote-allow-origins=http://127.0.0.1:{DEBUG_PORT}",
                f"--disable-extensions-except={extension_dir}",
                f"--load-extension={extension_dir}",
                "--no-sandbox",
                "--disable-setuid-sandbox",
                "--disable-dev-shm-usage",
                "--disable-gpu",
            ],
        )

        try:
            report["chromeVersion"] = http_json("/json/version").get("Browser")
        except Exception:
            pass

        # Capture errors from any extension page (popup/options) Playwright knows about.
        def wire_page(page):
            def on_console(msg):
                if msg.type == "error":
                    record(errors, "page.console", msg.text)
                elif msg.type == "warning":
                    record(warnings, "page.console", msg.text)
            page.on("console", on_console)
            page.on("pageerror", lambda exc: record(errors, "page.error", str(exc)))

        context.on("page", wire_page)
        for pg in context.pages:
            wire_page(pg)

        # Locate the extension service worker target via CDP (reliable for MV3).
        sw_target = None
        deadline = time.time() + LOAD_TIMEOUT_S
        while time.time() < deadline:
            try:
                targets = http_json("/json")
            except Exception:
                targets = []
            sw = [
                t for t in targets
                if t.get("type") == "service_worker"
                and str(t.get("url", "")).startswith("chrome-extension://")
            ]
            if sw:
                sw_target = sw[0]
                break
            time.sleep(0.5)

        ws = None
        if sw_target:
            report["loaded"] = True
            report["extensionId"] = sw_target["url"].split("/")[2]

            if websocket is not None and sw_target.get("webSocketDebuggerUrl"):
                try:
                    ws = websocket.create_connection(
                        sw_target["webSocketDebuggerUrl"], timeout=5
                    )
                    ws.send(json.dumps({"id": 1, "method": "Runtime.enable"}))
                    ws.send(json.dumps({"id": 2, "method": "Log.enable"}))
                except Exception as e:
                    record(warnings, "harness", f"Could not attach to service worker: {e}")
                    ws = None

        # Exercise the extension's own pages to wake the service worker and surface
        # errors that only occur once the extension actually runs.
        manifest = {}
        try:
            with open(os.path.join(extension_dir, "manifest.json")) as f:
                manifest = json.load(f)
        except Exception as e:
            record(errors, "manifest", f"Could not read manifest.json: {e}")

        ext_id = report["extensionId"]
        if ext_id:
            pages_to_open = []
            action = manifest.get("action") or manifest.get("browser_action") or {}
            if action.get("default_popup"):
                pages_to_open.append(action["default_popup"])
            if manifest.get("options_page"):
                pages_to_open.append(manifest["options_page"])
            options_ui = manifest.get("options_ui") or {}
            if options_ui.get("page"):
                pages_to_open.append(options_ui["page"])

            for rel in pages_to_open:
                try:
                    pg = context.new_page()
                    pg.goto(f"chrome-extension://{ext_id}/{rel}", timeout=8000)
                    pg.wait_for_timeout(1000)
                except Exception as e:
                    record(warnings, "harness", f"Could not open {rel}: {e}")

        # Drain service-worker CDP events for a short window.
        if ws is not None:
            end = time.time() + COLLECT_S
            ws.settimeout(1.0)
            while time.time() < end:
                try:
                    raw = ws.recv()
                except Exception:
                    continue
                if not raw:
                    continue
                try:
                    msg = json.loads(raw)
                except json.JSONDecodeError:
                    continue
                method = msg.get("method")
                params = msg.get("params", {})
                if method == "Runtime.exceptionThrown":
                    d = params.get("exceptionDetails", {})
                    text = (d.get("exception") or {}).get("description") or d.get("text")
                    record(errors, "service_worker.exception", text)
                elif method == "Runtime.consoleAPICalled":
                    text = " ".join(
                        str(a.get("value", a.get("description", "")))
                        for a in params.get("args", [])
                    )
                    t = params.get("type")
                    if t == "error":
                        record(errors, "service_worker.console", text)
                    elif t in ("warning", "warn"):
                        record(warnings, "service_worker.console", text)
                    else:
                        record(logs, f"service_worker.console.{t}", text)
                elif method == "Log.entryAdded":
                    entry = params.get("entry", {})
                    lvl = entry.get("level")
                    if lvl == "error":
                        record(errors, "service_worker.log", entry.get("text"))
                    elif lvl == "warning":
                        record(warnings, "service_worker.log", entry.get("text"))
            try:
                ws.close()
            except Exception:
                pass
        else:
            time.sleep(2)

        if not report["loaded"]:
            record(
                errors,
                "harness",
                "Extension service worker never registered — the extension failed to "
                "load. Check manifest.json (manifest_version, background.service_worker) "
                "and look for top-level errors in the service worker.",
            )

        context.close()

    with open(report_path, "w") as f:
        json.dump(report, f, indent=2)

    passed = report["loaded"] and not errors
    print(
        f"Extension verify {'PASSED' if passed else 'FAILED'} — "
        f"loaded={report['loaded']}, errors={len(errors)}, warnings={len(warnings)}"
    )
    for e in errors:
        print(f"  [error] ({e['source']}) {e['text']}")
    return 0 if passed else 1


if __name__ == "__main__":
    sys.exit(main())

File: spike/verify.mjs
// verify.mjs — the CRITICAL proof of the container plan.
//
// Loads a Chrome extension into a HEADED Chromium (running on the container's Xvfb display)
// and checks whether its MV3 service worker registers. This is the exact risk we set out to
// retire: MV3 extensions do NOT load under --headless=new, so a self-rolled image must
// provide a real X display. If this prints PASS for the migrated extension and FAIL for the
// unmigrated one, the container + browser design is proven.
//
// Node-native equivalent of the current Python browser_session.py: Playwright's
// launchPersistentContext exposes context.serviceWorkers() directly, so no manual CDP.
//
// Usage: node verify.mjs <extension-dir>   (exit 0 = SW registered, 1 = not)

import { chromium } from "playwright";
import { mkdtempSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { resolve } from "node:path";

const extDir = resolve(process.argv[2] ?? "fixtures/migrated-mv3");
const SW_TIMEOUT_MS = 15000;

function verdict(obj) {
  console.log(JSON.stringify(obj, null, 2));
}

const userDataDir = mkdtempSync(join(tmpdir(), "cft-profile-"));
let context;
const loadErrors = [];

try {
  // Exact Chrome flags carried over from browser_session.py — they transfer 1:1 to
  // Playwright-Node. Headed is mandatory; --no-sandbox + --disable-dev-shm-usage are the
  // two Docker-specific must-haves.
  context = await chromium.launchPersistentContext(userDataDir, {
    headless: false,
    timeout: 30000,
    args: [
      `--disable-extensions-except=${extDir}`,
      `--load-extension=${extDir}`,
      "--no-sandbox",
      "--disable-setuid-sandbox",
      "--disable-dev-shm-usage",
      "--disable-gpu",
      "--no-first-run",
      "--no-default-browser-check",
    ],
  });

  // Surface any page/worker errors that fire during load.
  context.on("weberror", (e) => loadErrors.push(String(e.error())));

  // MV3 background = a service worker. It may already be present, or arrive shortly after
  // launch. Check both.
  let [sw] = context.serviceWorkers();
  if (!sw) {
    sw = await context
      .waitForEvent("serviceworker", { timeout: SW_TIMEOUT_MS })
      .catch(() => null);
  }

  const browserVersion = context.browser()?.version() ?? "(persistent-context)";

  if (sw) {
    const extId = new URL(sw.url()).host;
    verdict({
      result: "PASS",
      extensionDir: extDir,
      serviceWorker: sw.url(),
      extensionId: extId,
      chromium: browserVersion,
      loadErrors,
    });
    await context.close();
    process.exit(0);
  } else {
    // No service worker registered within the window. For an MV2 (unmigrated) extension
    // this is the expected, correct failure — modern Chrome has no MV3 SW to register.
    verdict({
      result: "FAIL",
      reason: "no MV3 service worker registered within timeout",
      extensionDir: extDir,
      backgroundPages: context.backgroundPages().map((p) => p.url()),
      chromium: browserVersion,
      loadErrors,
    });
    await context.close();
    process.exit(1);
  }
} catch (err) {
  verdict({ result: "ERROR", extensionDir: extDir, error: String(err), loadErrors });
  try {
    await context?.close();
  } catch {}
  process.exit(2);
}
